src/Surfnet/StepupGateway/GatewayBundle/Sso2fa/CookieService.php

```diff
@@ -183,82 +159,46 @@ public function shouldSkip2faAuthentication(
-            $loa = $secondFactor->getLoaLevel($this->secondFactorTypeService);
-            if ($loa >= $requiredLoa) {
-                $this->logger->notice('Verified the current 2FA authentication can be given with the SSO on 2FA cookie');
-                $responseContext->saveSelectedSecondFactor($secondFactor);
-                return true;
-            }
-        }
-        return false;
+        $this->logger->notice('Verified the current 2FA authentication can be given with the SSO on 2FA cookie');
+        return true;
     }
 
-    public function getCookieFingerprint(Request $request): string
+    public function preconditionsAreMet(ResponseContext $responseContext): bool
     {
-        return $this->cookieHelper->fingerprint($request);
-    }
-
-    /**
-     * This method determines if an SSO on 2FA cookie should be created.
-     *
-     * The comments in the code block should give a good feel for what business rules
-     * are applied in this method.
-     *
-     * @param CookieValueInterface $ssoCookie           The SSO on 2FA cookie as read from the HTTP response
-     * @param float $loa                                The LoA that was requested for this authentication, used to
-     *                                                  compare to the LoA stored in the SSO cookie
-     * @param bool $wasAuthenticatedWithSsoOn2faCookie  Indicator if the currently running authentication was performed
-     *                                                  with the SSO on 2FA cookie
-     */
-    private function shouldAddCookie(
-        CookieValueInterface $ssoCookie,
-        bool $isCookieValid,
-        float $loa,
-        bool $wasAuthenticatedWithSsoOn2faCookie
-    ): bool {
-        // When the cookie is not yet set, was expired or was otherwise deemed invalid, we get a NullCookieValue
-        // back from the reader. Indicating there is no valid cookie present.
-        $cookieNotSet = $ssoCookie instanceof NullCookieValue;
-        // OR the existing cookie does exist, but the LoA stored in that cookie does not match the required LoA
-        $cookieDoesNotMeetLoaRequirement = ($ssoCookie instanceof CookieValue && !$ssoCookie->meetsRequiredLoa($loa));
-        if (!$ssoCookie instanceof NullCookieValue && $cookieDoesNotMeetLoaRequirement) {
+        if ($responseContext->isForceAuthn()) {
+            $this->logger->notice('Ignoring SSO on 2FA cookie when ForceAuthN is specified.');
+            return false;
+        }
+        $remoteSp = $this->getRemoteSp($responseContext);
+        // Test if the SP allows SSO on 2FA to take place (configured in MW config)
+        if (!$remoteSp->allowSsoOn2fa()) {
             $this->logger->notice(
                 sprintf(
-                    'Storing new SSO on 2FA cookie as LoA requirement (%d changed to %d) changed',
-                    $ssoCookie->getLoa(),
-                    $loa
+                    'Ignoring SSO on 2FA for SP: %s',
```

Comment:
My suggestion: 
'SSO on MFA is disabled by config for SP: %s'

Reply:
Agreed, but for sake of consistency I renamed MFA to 2FA